From 14c09e8eca430cc0dea2c23d4ef0f399b3a52626 Mon Sep 17 00:00:00 2001
From: Ryan Scott <ryan.gl.scott@gmail.com>
Date: Mon, 26 Nov 2018 12:59:30 -0500
Subject: [PATCH] Use autoconf to generate version numbers for libiserv and
 friends

Summary:
Currently, the version numbers for `libiserv`, `iserv`, and
`iserv-proxy` are hard-coded directly into their `.cabal` files.
These are easy to forget to update, and in fact, this has already
happened once (see #15866). Let's use `autoconf` to do this for us
so that it is not forgotten in the future.

Test Plan: ./validate

Reviewers: bgamari

Reviewed By: bgamari

Subscribers: rwbarton, erikd, carter

GHC Trac Issues: #15866

Differential Revision: https://phabricator.haskell.org/D5302
---
 .gitignore                                             |  3 +++
 configure.ac                                           |  2 +-
 .../libiserv/{libiserv.cabal => libiserv.cabal.in}     |  8 ++++++--
 .../{iserv-proxy.cabal => iserv-proxy.cabal.in}        |  8 ++++++--
 utils/iserv/{iserv.cabal => iserv.cabal.in}            | 10 +++++++---
 5 files changed, 23 insertions(+), 8 deletions(-)
 rename libraries/libiserv/{libiserv.cabal => libiserv.cabal.in} (80%)
 rename utils/iserv-proxy/{iserv-proxy.cabal => iserv-proxy.cabal.in} (91%)
 rename utils/iserv/{iserv.cabal => iserv.cabal.in} (81%)

diff --git a/configure.ac b/configure.ac
index 30241a5734..a5d2026434 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1298,7 +1298,7 @@ checkMake380() {
 checkMake380 make
 checkMake380 gmake
 
-AC_CONFIG_FILES([mk/config.mk mk/install.mk mk/project.mk rts/rts.cabal compiler/ghc.cabal ghc/ghc-bin.cabal utils/runghc/runghc.cabal utils/gen-dll/gen-dll.cabal libraries/ghc-boot/ghc-boot.cabal libraries/ghc-boot-th/ghc-boot-th.cabal libraries/ghci/ghci.cabal settings docs/users_guide/ghc_config.py docs/index.html libraries/prologue.txt distrib/configure.ac])
+AC_CONFIG_FILES([mk/config.mk mk/install.mk mk/project.mk rts/rts.cabal compiler/ghc.cabal ghc/ghc-bin.cabal utils/iserv/iserv.cabal utils/iserv-proxy/iserv-proxy.cabal utils/runghc/runghc.cabal utils/gen-dll/gen-dll.cabal libraries/ghc-boot/ghc-boot.cabal libraries/ghc-boot-th/ghc-boot-th.cabal libraries/ghci/ghci.cabal libraries/libiserv/libiserv.cabal settings docs/users_guide/ghc_config.py docs/index.html libraries/prologue.txt distrib/configure.ac])
 AC_OUTPUT
 [
 if test "$print_make_warning" = "true"; then
diff --git a/libraries/libiserv/libiserv.cabal b/libraries/libiserv/libiserv.cabal.in
similarity index 80%
rename from libraries/libiserv/libiserv.cabal
rename to libraries/libiserv/libiserv.cabal.in
index 43f1180073..386f26c034 100644
--- a/libraries/libiserv/libiserv.cabal
+++ b/libraries/libiserv/libiserv.cabal.in
@@ -1,5 +1,9 @@
+-- WARNING: libiserv.cabal is automatically generated from libiserv.cabal.in by
+-- ../../configure.  Make sure you are editing libiserv.cabal.in, not
+-- libiserv.cabal.
+
 Name: libiserv
-Version: 8.4
+Version: @ProjectVersionMunged@
 Copyright: XXX
 License: BSD3
 -- XXX License-File: LICENSE
@@ -25,7 +29,7 @@ Library
                    bytestring >= 0.10 && < 0.11,
                    containers >= 0.5 && < 0.6,
                    deepseq    >= 1.4 && < 1.5,
-                   ghci       == 8.4.*
+                   ghci       == @ProjectVersionMunged@
     if flag(network)
        Exposed-Modules: Remote.Message
                       , Remote.Slave
diff --git a/utils/iserv-proxy/iserv-proxy.cabal b/utils/iserv-proxy/iserv-proxy.cabal.in
similarity index 91%
rename from utils/iserv-proxy/iserv-proxy.cabal
rename to utils/iserv-proxy/iserv-proxy.cabal.in
index fa25c77794..2da65ae556 100644
--- a/utils/iserv-proxy/iserv-proxy.cabal
+++ b/utils/iserv-proxy/iserv-proxy.cabal.in
@@ -1,5 +1,9 @@
+-- WARNING: iserv-proxy.cabal is automatically generated from iserv-proxy.cabal.in by
+-- ../../configure.  Make sure you are editing iserv-proxy.cabal.in, not
+-- iserv-proxy.cabal.
+
 Name: iserv-proxy
-Version: 8.4
+Version: @ProjectVersion@
 Copyright: XXX
 License: BSD3
 -- XXX License-File: LICENSE
@@ -75,4 +79,4 @@ Executable iserv-proxy
                   directory  >= 1.3 && < 1.4,
                   network    >= 2.6,
                   filepath   >= 1.4 && < 1.5,
-                  libiserv   == 8.4
+                  libiserv   == @ProjectVersionMunged@
diff --git a/utils/iserv/iserv.cabal b/utils/iserv/iserv.cabal.in
similarity index 81%
rename from utils/iserv/iserv.cabal
rename to utils/iserv/iserv.cabal.in
index 768c0e0f83..d4234d4621 100644
--- a/utils/iserv/iserv.cabal
+++ b/utils/iserv/iserv.cabal.in
@@ -1,5 +1,9 @@
+-- WARNING: iserv.cabal is automatically generated from iserv.cabal.in by
+-- ../../configure.  Make sure you are editing iserv.cabal.in, not
+-- iserv.cabal.
+
 Name: iserv
-Version: 8.4
+Version: @ProjectVersion@
 Copyright: XXX
 License: BSD3
 -- XXX License-File: LICENSE
@@ -35,8 +39,8 @@ Executable iserv
                    bytestring >= 0.10 && < 0.11,
                    containers >= 0.5 && < 0.6,
                    deepseq    >= 1.4 && < 1.5,
-                   ghci       == 8.4.*,
-                   libiserv   == 8.4
+                   ghci       == @ProjectVersionMunged@,
+                   libiserv   == @ProjectVersionMunged@
 
     if os(windows)
         Cpp-Options: -DWINDOWS
-- 
2.24.1

